meltano_extension_sdk/extension.py

```diff
@@ -0,0 +1,123 @@
+"""Meltano extension SDK base class and supporting methods."""
+from __future__ import annotations
+
+import json
+import sys
+from abc import ABCMeta, abstractmethod
+from enum import Enum
+
+import structlog
+import yaml
+
+from meltano_extension_sdk import models
+
+
+class DescribeFormat(str, Enum):
+    text = "text"
+    json = "json"
+    yaml = "yaml"
+
+
+class ExtensionBase(metaclass=ABCMeta):
+    """Basic extension interface that must be implemented by all extensions."""
+
+    def pre_invoke(self):
+        """Called before the extension is invoked."""
+        pass
+
+    def initialize(self, force: bool = False):
+        """Initialize the extension.
+
+        This method is called on-demand by the user to initialize the extension.
+        Extensions are not required to implement this method, and may no-op.
+
+        Args:
+            force: If True, force initialization.
+        """
+        pass
+
+    @abstractmethod
+    def invoke(self, command_name: str | None, *command_args) -> None:
+        """Invoke method.
+
+        This method is called when the extension is invoked.
+        """
+        pass
+
+    def post_invoke(self):
+        """Called after the extension is invoked."""
+        pass
+
+    @abstractmethod
+    def describe(self) -> models.Describe:
+        """Describe method.
+
+        This method should describe what commands & capabilities the extension provides.
+
+        Returns:
+            Description: A description of the extension.
+        """
+        pass
+
+    def describe_formatted(
+        self, output_format: DescribeFormat = DescribeFormat.text
+    ) -> str:
+        """Return a formatted description of the extensions commands and capabilities.
+
+        Args:
+            output_format: The output format to use.
+
+        Returns:
+            str: The formatted description.
+        """
+        if output_format == DescribeFormat.text:
```

Comment:
I wonder if we shouldn't have a 4th output format called `meltano` that actually formats the yaml output to match what we'd want in the meltano.yaml e.g.:

```yaml
$ poetry run airflow_extension describe --format=meltano
executable: airflow_invoker
commands:
  describe:
    executable: airflow_extension
    args: describe
  initialize:
    executable: airflow_extension
    args: initialize
  invoke:
    executable: airflow_extension
    args: invoke
```